libyul/SideEffects.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Specific AST walkers that collect semantical facts.
+ */
+
+#include <libyul/optimiser/Semantics.h>
```

Comment:
I'd consider renaming and restructuring the files a bit - it's getting a bit confusing...
This here should be ``libyul/SideEffectsCollector.cpp`` and maybe a ``libyul/SideEffectsCollector.h`` should be extracted from ``libyul/optimiser/Semantics.h``?

At the moment this is called ``SideEffects.cpp``, but implements nothing declared in ``SideEffects.h`` and one needs to look for the declaration of ``SideEffectsCollector`` until one finds it in ``Semantics.h``...

Reply:
The reason is that this file is not in the ``CMakeLists.txt``, so it won't be built (it's not used in this PR yet it seems)...